test/replication-luatest/gh_6126_auto_rejoin_test.lua

```diff
@@ -0,0 +1,119 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local fio = require('fio')
+
+local g = t.group('gh_6126')
+
+g.before_each(function(cg)
+    cg.cluster = cluster:new({})
+    local box_cfg = {
+        replication = server.build_instance_uri('master'),
+    }
+    cg.replica = cg.cluster:build_and_add_server(
+        {alias = 'replica', box_cfg = box_cfg})
+    cg.master = cg.cluster:build_and_add_server(
+        {alias = 'master', box_cfg = box_cfg})
+end)
+
+g.after_each(function(cg)
+    cg.cluster:drop()
+end)
+
+g.test_auto_rejoin = function(cg)
+    -- Test that joining replica tolerates ER_READONLY errors from master
+    cg.master:start()
+    cg.master:exec(function() box.cfg{read_only = true} end)
+    cg.replica:start({wait_for_readiness = false})
+    local logfile = fio.pathjoin(cg.replica.workdir, 'replica.log')
+    t.helpers.retrying({}, function()
+        t.assert(cg.replica:grep_log('ER_READONLY', nil,
+            {filename = logfile}), 'Can\'t modify data on a read-only'..
+            'instance - box.cfg.read_only is true')
+    end)
+    t.assert(g.master:exec(function()
+        return box.space._cluster:count() == 1
+    end),  'No join while master is read-only')
+
+    -- Test that replica joins after master became writeable
+    cg.master:exec(function() box.cfg{read_only = false} end)
+    cg.replica:wait_for_readiness()
+    t.helpers.retrying({}, function()
+        t.assert(g.master:exec(function()
+            return box.space._cluster:count() == 2
+        end), 'Join after master became writeable')
+        cg.replica:assert_follows_upstream(1)
+    end)
+end
+
+g.test_replicaset_needs_rejoin = function(cg)
+    -- Instance 1
+    cg.master:start()
+
+    -- Instance 2
+    cg.replica:start()
+    cg.replica:stop()
+
+    -- Bump vclock on instance 1
+    cg.master:exec(function()
+        box.space._schema:replace{'something'}
+        box.snapshot()
+    end)
+    cg.master:stop()
+
+    -- In 1st instance data dir: rm *.xlog
+    local dirname = fio.pathjoin(cg.master.workdir)
+    local xlogs = fio.glob(dirname .. "/*.xlog")
+    fio.unlink(xlogs[#xlogs])
+
+    -- Restart all instanses
+    cg.master:start()
+    cg.replica:start()
+    local logfile = fio.pathjoin(cg.replica.workdir, 'replica.log')
+    t.helpers.retrying({}, function()
+        t.assert(cg.replica:grep_log('replica is too old', nil,
+                {filename = logfile}), 'initiating rebootstrap')
+    end)
+end
+
+g.test_access_denied = function(cg)
```

Comment:
It seems a proper test case would be something like:
Start two instances, anonymous and normal. Configure their instance_uuid so that anonymous instance has a smaller uuid than a normal one (so that it's chosen a bootstrap leader by some third node). Let them finish bootstrap. Set normal instance to read-only.

Now try to add third instance. It should fail with ER_UNSUPPORTED. because it will chose an anonymous replica as its bootstrap master.
Let it retry, then set normal instance to `box.cfg{read_only=false}`.

Now 3rd instance should find another node to boot off (normal one) and boot successfully.

I tried similar test locally and it ended up in a segfault on a third node:
```
2022-07-06 17:01:42.276 [27202] main/103/interactive I> connecting to 1 replicas
2022-07-06 17:01:42.277 [27202] main/110/applier/ I> remote master dac002d9-adcb-4845-b7f5-a5e8326ec2d4 at [::1]:3301 running Tarantool 2.11.0
2022-07-06 17:01:42.277 [27202] main/103/interactive I> connected to 1 replicas
2022-07-06 17:01:42.277 [27202] main/103/interactive I> bootstrapping replica from dac002d9-adcb-4845-b7f5-a5e8326ec2d4 at [::1]:3301
2022-07-06 17:01:42.278 [27202] main/110/applier/ I> can't read row
2022-07-06 17:01:42.278 [27202] main/110/applier/ box.cc:3050 E> ER_UNSUPPORTED: Anonymous replica does not support registration of non-anonymous nodes.
2022-07-06 17:01:42.278 [27202] main/103/interactive I> join attempt has failed, will retry every 1.00 second
2022-07-06 17:01:43.278 [27202] main/103/interactive I> connecting to 1 replicas
2022-07-06 17:01:43.279 [27202] main/111/applier/ I> remote master dac002d9-adcb-4845-b7f5-a5e8326ec2d4 at [::1]:3301 running Tarantool 2.11.0
2022-07-06 17:01:43.279 [27202] main/103/interactive I> connected to 1 replicas
Segmentation fault
  code: SEGV_MAPERR
  addr: 0x8
  context: 0x11617fb68
  siginfo: 0x11617fb00
Current time: 1657116103
Please file a bug at http://github.com/tarantool/tarantool/issues
Attempting backtrace... Note: since the server has already crashed, 
this may fail as well
#1  0x10a1ff3ab in crash_collect+187
#2  0x10a1ff25c in crash_signal_cb+60
#3  0x7fff2038cd7d in _sigtramp+29
#4  0x10a0c8a2e in replica_on_applier_state_f(trigger*, void*)+62
#5  0x10a2317ac in trigger_run+140
#6  0x10a0cf16d in trigger_run_xc(rlist*, void*)+29
#7  0x10a0cd413 in applier_set_state(applier*, applier_state)+131
#8  0x10a0d063b in applier_disconnect(applier*, applier_state)+27
#9  0x10a0ce69a in applier_f(__va_list_tag*)+2554
#10 0x109f7a3da in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+26
#11 0x10a209cc7 in fiber_loop+183
#12 0x10a57e6f2 in coro_init+114
Abort trap: 6
```

Now at least something is tested.